executor/src/witgen/jit/block_machine_processor.rs

```diff
@@ -160,7 +267,16 @@ impl<T: FieldElement> FixedEvaluator<T> for &BlockMachineProcessor<'_, T> {
     fn evaluate(&self, var: &AlgebraicReference, row_offset: i32) -> Option<T> {
         assert!(var.is_fixed());
         let values = self.fixed_data.fixed_cols[&var.poly_id].values_max_size();
-        let row = (row_offset + var.next as i32 + values.len() as i32) as usize % values.len();
+
+        // By assumption of the block machine, all fixed columns are cyclic with a period of <block_size>.
+        // An exception might be the first and last row.
+        assert!(row_offset >= -1);
```

Comment:
```suggestion
        assert!(row_offset >= -self.block_size);
```
?

Reply:
Also what was wrong with the previous code?

Reply:
Why `-block_size`? We should access at least `-1` with the current row range. We can relax this later if needed, but I think it's good to be strict at this time.

The idea here is that we assume that in a block machine, all fixed columns are cyclic *except* the first and last row. This code makes sure that we never access rows 0 and -1. 

I added some more assertions and a comment.

Reply:
Ok, now I understand. The main problem is that we don't really check that property anywhere and I also don't think it is documented in the machine extractor or the `try_new` function of the block machine.